File: server/src/main/java/com/bomberman/game/Game.java
package com.bomberman.game;

import com.bomberman.entities.Movement;
import com.bomberman.entities.Position;
import com.bomberman.entities.mapobject.Bomb;
import com.bomberman.entities.mapobject.BombExplosion;
import com.bomberman.entities.mapobject.MapObject;
import com.bomberman.entities.mapobject.movable.Player;
import com.bomberman.exception.PlayerNotFoundException;
import com.bomberman.factories.LevelFactory;
import com.bomberman.factories.PlayerFactory;
import com.bomberman.utils.Collisions;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class Game {

  private final PlayerFactory playerFactory;
  private final GameState gameState;

  public Game() {
    playerFactory = new PlayerFactory();
    final LevelFactory levelFactory = new LevelFactory();
    gameState = levelFactory.createLevel();
  }

  public void addPlayer(final String id) {
    final var newPlayer = playerFactory.createPlayer(id, gameState);

    gameState.getPlayers().add(newPlayer);
  }

  public void removePlayer(final String id) {
    gameState.getPlayers().removeIf(x -> x.getId().equals(id));
  }

  public void movePlayer(final String playerId, final Movement movement) {
    final var player =
        gameState.getPlayers().stream()
            .filter($player -> $player.getId().equals(playerId))
            .findFirst()
            .orElseThrow(() -> new PlayerNotFoundException(playerId));

    if (player.isDead()) {
      return;
    }

    final float movementSpeed = player.getSpeed();
    final var position = movement.getDirection().buildPosition(player.getPosition(), movementSpeed);
    player.move(position);

    player.addStateTime(movement.getDt());

    final var explosion =
        gameState.getBombExplosions().stream()
            .filter($explosion -> Collisions.willCollide(player, player.getPosition(), $explosion))
            .findFirst();

    explosion.ifPresent($explosion -> updateDeathsAndScore($explosion.getPlayerId(), player));
  }

  private void updateDeathsAndScore(final String killerId, final Player player) {
    final var killerOptional =
        gameState.getPlayers().stream()
            .filter($player -> $player.getId().equals(killerId))
            .findFirst();
    killerOptional.ifPresent($killer -> updateDeathsAndScore($killer, player));
  }

  public Optional<Bomb> placeBomb(final String playerId) {
    final var player = getPlayer(playerId);
    if (!player.canPlaceBomb()) {
      return Optional.empty();
    }

    final var position = Position.round(player.getPosition());
    final var bomb =
        new Bomb(position, player.getBombDurationInSeconds(), player, gameState.getPlayers());
    if (Collisions.willCollide(bomb, position, new ArrayList<>(gameState.getBombs()))) {
      return Optional.empty();
    }
    gameState.getBombs().add(bomb);
    player.incrementPlacedBombCount();
    return Optional.of(bomb);
  }

  public void explodeBomb(final Bomb bomb, final String bombId, final String playerId) {
    if (!gameState.getBombs().contains(bomb)) {
      return;
    }
    final var player = getPlayer(playerId);
    player.decrementPlacedBombCount();
    final var bombRange = bomb.getBombRange();
    gameState.getBombs().remove(bomb);
    final List<Bomb> bombsInExplosionRange =
        gameState.getBombs().stream()
            .filter($bomb -> Collisions.isInExplosionRange(bombRange, bomb, $bomb, gameState))
            .collect(Collectors.toList());

    gameState.getBombExplosions().addAll(BombExplosion.of(bomb, bombId, playerId, gameState));
    bombsInExplosionRange.forEach(System.out::println);
    bombsInExplosionRange.forEach($bomb -> explodeBomb($bomb, bombId, playerId));
  }

  public void checkKilled(final Bomb bomb) {
    final List<Player> playersInExplosionRange =
        gameState.getPlayers().stream()
            .filter(
                $player ->
                    Collisions.willCollide(
                        $player,
                        $player.getPosition(),
                        new ArrayList<MapObject>(gameState.getBombExplosions())))
            .collect(Collectors.toList());

    final var killer =
        gameState.getPlayers().stream()
            .filter($player -> $player.getId().equals(bomb.getPlayerId()))
            .findFirst()
            .orElseThrow();

    playersInExplosionRange.forEach($player -> updateDeathsAndScore(killer, $player));
  }

  private void updateDeathsAndScore(final Player killer, final Player killed) {
    if (killed.isDead()) {
      return;
    }

    if (Objects.equals(killer, killed)) {
      killer.killedHimself();
      killer.setDead();
      System.out.println(String.format("Player %s killed himself", killer.getId()));
    } else {
      killer.killedEnemy();
      killed.setDead();
      System.out.println(
          String.format("Player %s died from player %s", killed.getId(), killer.getId()));
    }
  }

  public void removeExplosions(final Bomb bomb) {
    final var explosions =
        gameState.getBombExplosions().stream()
            .filter($explosion -> $explosion.getBombId().equals(bomb.getBombId()))
            .collect(Collectors.toList());

    final var bricksInExplosion =
        gameState.getBricks().stream()
            .filter(
                $brick ->
                    explosions.stream()
                        .anyMatch(
                            $explosion ->
                                Objects.equals($brick.getPosition(), $explosion.getPosition())))
            .collect(Collectors.toList());

    gameState.getBricks().removeAll(bricksInExplosion);
    gameState.getBombExplosions().removeAll(explosions);
  }

  private Player getPlayer(final String playerId) {
    return gameState.getPlayers().stream()
        .filter(x -> x.getId().equals(playerId))
        .findFirst()
        .orElse(null);
  }

  public GameState getGameState() {
    return gameState;
  }

  public void respawnPlayer(final String playerId) {
    final var playerOptional =
        gameState.getPlayers().stream().filter(x -> x.getId().equals(playerId)).findFirst();

    playerOptional.ifPresent(Player::respawn);
  }

  public void increasePlayerSpeed(final String playerId) {
    final var playerOptional =
        gameState.getPlayers().stream().filter(x -> x.getId().equals(playerId)).findFirst();

    playerOptional.ifPresent(Player::increasePlayerSpeed);
  }

  public void decreasePlayerSpeed(final String playerId) {
    final var playerOptional =
        gameState.getPlayers().stream().filter(x -> x.getId().equals(playerId)).findFirst();

    playerOptional.ifPresent(Player::decreasePlayerSpeed);
  }

  public void increaseBombRange(final String playerId) {
    final var playerOptional =
        gameState.getPlayers().stream().filter(x -> x.getId().equals(playerId)).findFirst();

    playerOptional.ifPresent(Player::increaseBombRange);
  }

  public void decreaseBombRange(final String playerId) {
    final var playerOptional =
        gameState.getPlayers().stream().filter(x -> x.getId().equals(playerId)).findFirst();

    playerOptional.ifPresent(Player::decreaseBombRange);
  }

  public void increaseMaxBombCount(final String playerId) {
    final var playerOptional =
        gameState.getPlayers().stream().filter(x -> x.getId().equals(playerId)).findFirst();

    playerOptional.ifPresent(Player::increaseMaxBombCount);
  }
}

File: server/src/main/kotlin/com/bomberman/game/Game.kt
package com.bomberman.game

import com.bomberman.PlayerAlreadyExistsException
import com.bomberman.entity.Direction
import com.bomberman.entity.mapobject.Bomb
import com.bomberman.entity.mapobject.Brick
import com.bomberman.entity.mapobject.movable.Player
import com.bomberman.entity.mapobject.pickup.BombPickup
import com.bomberman.entity.mapobject.pickup.Pickup
import com.bomberman.factory.LevelFactory
import com.bomberman.factory.PickupFactory
import com.bomberman.factory.PlayerFactory

class Game(
    private val playerFactory: PlayerFactory,
    private val pickupFactory: PickupFactory,
    val gameState: GameState
) {

  constructor() : this(LevelFactory().createLevel())
  constructor(gameState: GameState) : this(PlayerFactory(), PickupFactory(), gameState)

  fun addPlayer(playerId: String) {
    if (playerExists(playerId)) {
      throw PlayerAlreadyExistsException(playerId)
    }

    val newPlayer = playerFactory.createPlayer(playerId, gameState)
    gameState.players.add(newPlayer)
  }

  fun playerExists(id: String): Boolean =
      gameState.players.any { it.id == id }

  fun removePlayer(id: String) {
    gameState.players.removeIf { it.id == id }
  }

  private fun getPlayer(playerId: String): Player =
      gameState.players.first { it.id == playerId }

  fun movePlayer(playerId: String, direction: Direction) {
    val player = getPlayer(playerId)

    if (player.dead) {
      return
    }

    val movementSpeed = player.speed
    val position = direction.buildPosition(player.position, movementSpeed)
    player.move(position)

    val collidePickup = player.firstPickupCollision(position, gameState.pickups)
    collidePickup?.applyAndRemove(player)

    val explosion = gameState.bombExplosions.firstOrNull { player.willCollide(player.position, it) }

    explosion?.let { updateDeathsAndScore(it.playerId, player) }
  }

  private fun updateDeathsAndScore(killerId: String, player: Player) {
    val killer = gameState.players.firstOrNull { it.id == killerId }
    killer?.updateDeathsAndScore(player)
  }

  fun placeBomb(playerId: String): Bomb? {
    val player = getPlayer(playerId)
    if (!player.canPlaceBomb()) {
      return null
    }

    val position = player.position.round()
    val bomb = Bomb(position, player.bombDurationInSeconds, player, gameState.players)
    if (bomb.willCollide(position, gameState.bombs)) {
      return null
    }
    gameState.bombs += bomb
    player.incrementPlacedBombCount()
    return bomb
  }

  fun explodeBomb(bomb: Bomb, bombId: String, playerId: String) {
    if (bomb !in gameState.bombs) {
      return
    }
    val player = getPlayer(playerId)
    player.decrementPlacedBombCount()
    gameState.bombs -= bomb

    val bombsInExplosionsRange = gameState.bombs
        .filter { it.isInExplosionRange(bomb.bombRange, bomb, gameState.collideObjects) }

    gameState.bombExplosions += bomb.bombExplosions(bombId, playerId, gameState)
    bombsInExplosionsRange.forEach { println(it) }
    bombsInExplosionsRange.forEach { explodeBomb(it, bombId, playerId) }
  }

  fun respawnPlayer(playerId: String) {
    getPlayer(playerId).respawn()
  }

  fun checkKilled(bomb: Bomb) {
    val playersInExplosionRange = gameState.players.filter {
      it.willCollide(it.position, gameState.bombExplosions)
    }
    val killer = getPlayer(bomb.playerId)
    playersInExplosionRange.forEach { killer.updateDeathsAndScore(it) }
  }

  private fun Player.updateDeathsAndScore(killedPlayer: Player) {
    if (killedPlayer.dead) {
      return
    }
    if (this.id == killedPlayer.id) {
      this.killedHimself()
      this.dead = true
      println("Player ${this.id} killed himself")
    } else {
      this.killedEnemy()
      killedPlayer.dead = true
      println("Player ${killedPlayer.id} died from player ${this.id}")
    }
  }

  fun removeExplosions(bomb: Bomb) {
    val explosions = gameState.bombExplosions.filter { it.bombId == bomb.bombId }

    val bricksInExplosion = gameState.bricks.filter { brick ->
      explosions.any { explosion ->
        brick.position == explosion.position
      }
    }

    bricksInExplosion
        .map { it.createPickupAndRemoveBrick() }
        .forEach { gameState.pickups += it }

    gameState.bombExplosions -= explosions
  }

  fun increasePlayerSpeed(playerId: String) {
    getPlayer(playerId).increasePlayerSpeed()
  }

  fun decreasePlayerSpeed(playerId: String) {
    getPlayer(playerId).decreasePlayerSpeed()
  }

  fun increaseBombRange(playerId: String) {
    getPlayer(playerId).increaseBombRange()
  }

  fun decreaseBombRange(playerId: String) {
    getPlayer(playerId).decreaseBombRange()
  }

  fun increaseMaxBombCount(playerId: String) {
    getPlayer(playerId).increaseMaxBombCount()
  }

  private fun Brick.createPickupAndRemoveBrick(): BombPickup {
    gameState.bricks.remove(this)
    return pickupFactory.createBombPickup(this)
  }

  private fun Pickup.applyAndRemove(player: Player) {
    this.apply(player)
    if (this is BombPickup) {
      gameState.pickups.remove(this)
    }
  }

}